bor-mainnet: bridge nil-ptr

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x178183d]

goroutine 78 [running]:
github.com/ledgerwatch/erigon/polygon/bridge.(*Bridge).GetEvents(0x0, {0x31e2140, 0x552fd00}, 0x37d4870)
	github.com/ledgerwatch/erigon/polygon/bridge/bridge.go:163 +0x3d
github.com/ledgerwatch/erigon/polygon/bor.(*Bor).CommitStates(0xc00182a780, 0xc4adb8b7a0?, 0xc4a8dd8dc8?, {{0x3201550?, 0xc438062d00?}, {0x3206ed0?, 0xc00182a780?}}, 0xc4e0e5be00)
	github.com/ledgerwatch/erigon/polygon/bor/bor.go:1473 +0x9a
github.com/ledgerwatch/erigon/polygon/bor.(*Bor).Finalize(0xc00182a780, 0x0?, 0xc4a8dd8dc8, 0xc4adb8b7a0, {0x805cedf6?, 0xc4e40c5010?, 0x0?}, {0x0, 0x0, 0x0}, ...)
	github.com/ledgerwatch/erigon/polygon/bor/bor.go:1010 +0x312
github.com/ledgerwatch/erigon/cmd/state/exec3.(*Worker).RunTxTaskNoLock(0xc4da25c488, 0xc4e40c2a88)
	github.com/ledgerwatch/erigon/cmd/state/exec3/state.go:210 +0x864
github.com/ledgerwatch/erigon/eth/stagedsync.ExecV3({_, _}, _, {_, _}, _, {{0x31f9a70, 0xc2e862a990}, 0x40000000, {0x1, ...}, ...}, ...)
	github.com/ledgerwatch/erigon/eth/stagedsync/exec3.go:770 +0x2c69
github.com/ledgerwatch/erigon/eth/stagedsync.ExecBlockV3(_, {_, _}, {{_, _}, {_, _}, _}, _, {0x31e2620, ...}, ...)
	github.com/ledgerwatch/erigon/eth/stagedsync/stage_execute.go:275 +0x1ea
github.com/ledgerwatch/erigon/eth/stagedsync.SpawnExecuteBlocksStage(_, {_, _}, {{_, _}, {_, _}, _}, _, {0x31e2620, ...}, ...)
	github.com/ledgerwatch/erigon/eth/stagedsync/stage_execute.go:397 +0x106
github.com/ledgerwatch/erigon/eth/stagedsync.DefaultStages.func19(0x9?, 0x0?, {0x31dc570?, 0xc1daedcc60?}, {{0x0, 0x0}, {0x0, 0x0}, 0x0}, {0x31f8668, ...})
	github.com/ledgerwatch/erigon/eth/stagedsync/default_stages.go:118 +0xf0
github.com/ledgerwatch/erigon/eth/stagedsync.(*Sync).runStage(0xc1daedcc60, 0xc2106ff900, {0x31f9a70, 0xc2e862a990}, {{0x0, 0x0}, {0x0, 0x0}, 0x0}, 0x1, ...)
	github.com/ledgerwatch/erigon/eth/stagedsync/sync.go:527 +0x190
github.com/ledgerwatch/erigon/eth/stagedsync.(*Sync).Run(0xc1daedcc60, {0x31f9a70, 0xc2e862a990}, {{0x0, 0x0}, {0x0, 0x0}, 0x0}, 0xf8?, 0x1)
	github.com/ledgerwatch/erigon/eth/stagedsync/sync.go:397 +0x2ad
github.com/ledgerwatch/erigon/turbo/stages.ProcessFrozenBlocks({0x31e2620, 0xc000ced5e0}, {0x31f9a70, 0xc2e862a990}, {0x321dd60, 0xc2e43f8180}, 0xc1daedcc60)
	github.com/ledgerwatch/erigon/turbo/stages/stageloop.go:128 +0x97
github.com/ledgerwatch/erigon/turbo/stages.StageLoop({0x31e2620, 0xc000ced5e0}, {0x31f9a70, 0xc2e862a990}, 0xc1daedcc60, 0xc2e0e83c00, 0x3206e28?, 0x0, {0x31f8668, 0xc000a5e2e0}, ...)
	github.com/ledgerwatch/erigon/turbo/stages/stageloop.go:63 +0xbf
created by github.com/ledgerwatch/erigon/eth.(*Ethereum).Start in goroutine 1
	github.com/ledgerwatch/erigon/eth/backend.go:1571 +0x746
```

flags:
```
 --torrent.download.rate 10G --torrent.download.slots=9_000 --datadir="/erigon-data" --chain=bor-mainnet --bor.heimdall https://heimdall-api.polygon.technology --txpool.disable --sync.loop.block.limit=9_000 --batchSize=1g --metrics --metrics.addr=0.0.0.0 --metrics.port="8999" --pprof --pprof.port=6160 --torrent.verbosity=3 --log.console.verbosity=4 --no-downloader --http.api=eth,erigon,ots,web3,net,debug,trace,txpool
``

--- a/polygon/bor/bor.go
+++ b/polygon/bor/bor.go
@@ -1469,7 +1469,7 @@ func (c *Bor) CommitStates(
 ) error {
 	blockNum := header.Number.Uint64()
 
-	if c.polygonBridge != nil {
+	if c.polygonBridge.(*bridge.Bridge) != nil {
 		events, err := c.polygonBridge.GetEvents(c.execCtx, blockNum)
 		if err != nil {
 			return err

--- a/polygon/bridge/bridge.go
+++ b/polygon/bridge/bridge.go
@@ -57,6 +57,7 @@ func (b *Bridge) Run(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
+	defer b.Close()
 
 	// get last known sync ID
 	lastEventID, err := b.store.GetLatestEventID(ctx)
@@ -110,7 +111,7 @@ func (b *Bridge) ProcessNewBlocks(ctx context.Context, blocks []*types.Block) er
 	eventMap := make(map[uint64]uint64)
 	for _, block := range blocks {
 		// check if block is start of span
-		if b.isSprintStart(block.NumberU64()) {
+		if !b.isSprintStart(block.NumberU64()) {
 			continue
 		}
 
@@ -164,12 +165,16 @@ func (b *Bridge) GetEvents(ctx context.Context, blockNum uint64) ([]*types.Messa
 		return nil, err
 	}
 
+	if end == 0 { // exception for tip processing
+		end = b.lastProcessedEventID
+	}
+
 	b.log.Debug("got map", "blockNum", blockNum, "start", start, "end", end)
 
-	eventsRaw := make([]*types.Message, end-start+1)
+	eventsRaw := make([]*types.Message, 0, end-start+1)
 
 	// get events from DB
-	events, err := b.store.GetEvents(ctx, start, end)
+	events, err := b.store.GetEvents(ctx, start+1, end+1)
 	if err != nil {
 		return nil, err
 	}
@@ -192,6 +197,7 @@ func (b *Bridge) GetEvents(ctx context.Context, blockNum uint64) ([]*types.Messa
 
 		eventsRaw = append(eventsRaw, &msg)
 	}
+
 	return eventsRaw, nil
 }
 

--- a/polygon/bridge/bridge_test.go
+++ b/polygon/bridge/bridge_test.go
@@ -1,0 +1,251 @@
+package bridge_test
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+
+	"github.com/ledgerwatch/erigon-lib/common/hexutil"
+	"github.com/ledgerwatch/erigon-lib/log/v3"
+	"github.com/ledgerwatch/erigon/accounts/abi"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/erigon/polygon/bor"
+	"github.com/ledgerwatch/erigon/polygon/bor/borcfg"
+	"github.com/ledgerwatch/erigon/polygon/bridge"
+	"github.com/ledgerwatch/erigon/polygon/heimdall"
+	"github.com/ledgerwatch/erigon/polygon/polygoncommon"
+	"github.com/ledgerwatch/erigon/rlp"
+	"github.com/ledgerwatch/erigon/turbo/testlog"
+)
+
+func setup(t *testing.T, abi abi.ABI) (*heimdall.MockHeimdallClient, *bridge.Bridge) {
+	ctrl := gomock.NewController(t)
+	logger := testlog.Logger(t, log.LvlDebug)
+	borConfig := borcfg.BorConfig{
+		Sprint:                map[string]uint64{"0": 2},
+		StateReceiverContract: "0x0000000000000000000000000000000000001001",
+	}
+
+	heimdallClient := heimdall.NewMockHeimdallClient(ctrl)
+	polygonBridgeDB := polygoncommon.NewDatabase(t.TempDir(), logger)
+	store := bridge.NewStore(polygonBridgeDB)
+	b := bridge.NewBridge(store, logger, &borConfig, heimdallClient.FetchStateSyncEvents, abi)
+
+	return heimdallClient, b
+}
+
+func getBlocks(t *testing.T, numBlocks int) []*types.Block {
+	// Feed in new blocks
+	rawBlocks := make([]*types.RawBlock, 0, numBlocks)
+
+	for i := 0; i < numBlocks; i++ {
+		rawBlocks = append(rawBlocks, &types.RawBlock{
+			Header: &types.Header{
+				Number: big.NewInt(int64(i)),
+				Time:   uint64(50 * (i + 1)),
+			},
+			Body: &types.RawBody{},
+		})
+	}
+
+	blocks := make([]*types.Block, len(rawBlocks))
+
+	for i, rawBlock := range rawBlocks {
+		b, err := rawBlock.AsBlock()
+		require.NoError(t, err)
+
+		blocks[i] = b
+	}
+
+	return blocks
+}
+
+func TestBridge(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	stateReceiverABI := bor.GenesisContractStateReceiverABI()
+	heimdallClient, b := setup(t, stateReceiverABI)
+
+	event1 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      1,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x01"),
+		},
+		Time: time.Unix(50, 0), // block 2
+	}
+	event2 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      2,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x02"),
+		},
+		Time: time.Unix(100, 0), // block 2
+	}
+	event3 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      3,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x03"),
+		},
+		Time: time.Unix(200, 0), // block 4
+	}
+
+	events := []*heimdall.EventRecordWithTime{event1, event2, event3}
+
+	heimdallClient.EXPECT().FetchStateSyncEvents(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(events, nil).Times(1)
+	heimdallClient.EXPECT().FetchStateSyncEvents(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return([]*heimdall.EventRecordWithTime{}, nil).AnyTimes()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func(bridge bridge.Service) {
+		defer wg.Done()
+
+		err := bridge.Run(ctx)
+		if err != nil {
+			if !errors.Is(err, ctx.Err()) {
+				t.Error(err)
+			}
+
+			return
+		}
+	}(b)
+
+	err := b.Synchronize(ctx, &types.Header{Number: big.NewInt(100)}) // hack to wait for b.ready
+	require.NoError(t, err)
+
+	blocks := getBlocks(t, 5)
+
+	err = b.ProcessNewBlocks(ctx, blocks)
+	require.NoError(t, err)
+
+	res, err := b.GetEvents(ctx, 2)
+	require.NoError(t, err)
+
+	event1Data, err := event1.Pack(stateReceiverABI)
+	require.NoError(t, err)
+
+	event2Data, err := event2.Pack(stateReceiverABI)
+	require.NoError(t, err)
+
+	require.Equal(t, 2, len(res))                             // have first two events
+	require.Equal(t, event1Data, rlp.RawValue(res[0].Data())) // check data fields
+	require.Equal(t, event2Data, rlp.RawValue(res[1].Data()))
+
+	res, err = b.GetEvents(ctx, 4)
+	require.NoError(t, err)
+
+	event3Data, err := event3.Pack(stateReceiverABI)
+	require.NoError(t, err)
+
+	require.Equal(t, 1, len(res))
+	require.Equal(t, event3Data, rlp.RawValue(res[0].Data()))
+
+	// get non-sprint block
+	_, err = b.GetEvents(ctx, 1)
+	require.Error(t, err)
+
+	_, err = b.GetEvents(ctx, 3)
+	require.Error(t, err)
+
+	// check block 0
+	_, err = b.GetEvents(ctx, 0)
+	require.Error(t, err)
+
+	cancel()
+	wg.Wait()
+}
+
+func TestBridge_Unwind(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	stateReceiverABI := bor.GenesisContractStateReceiverABI()
+	heimdallClient, b := setup(t, stateReceiverABI)
+
+	event1 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      1,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x01"),
+		},
+		Time: time.Unix(50, 0), // block 2
+	}
+	event2 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      2,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x02"),
+		},
+		Time: time.Unix(100, 0), // block 2
+	}
+	event3 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      3,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x03"),
+		},
+		Time: time.Unix(200, 0), // block 4
+	}
+	event4 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      4,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x03"),
+		},
+		Time: time.Unix(300, 0), // block 6
+	}
+
+	events := []*heimdall.EventRecordWithTime{event1, event2, event3, event4}
+
+	heimdallClient.EXPECT().FetchStateSyncEvents(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(events, nil).Times(1)
+	heimdallClient.EXPECT().FetchStateSyncEvents(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return([]*heimdall.EventRecordWithTime{}, nil).AnyTimes()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func(bridge bridge.Service) {
+		defer wg.Done()
+
+		err := bridge.Run(ctx)
+		if err != nil {
+			if !errors.Is(err, ctx.Err()) {
+				t.Error(err)
+			}
+
+			return
+		}
+	}(b)
+
+	err := b.Synchronize(ctx, &types.Header{Number: big.NewInt(100)}) // hack to wait for b.ready
+	require.NoError(t, err)
+
+	blocks := getBlocks(t, 8)
+
+	err = b.ProcessNewBlocks(ctx, blocks)
+	require.NoError(t, err)
+
+	event3Data, err := event3.Pack(stateReceiverABI)
+	require.NoError(t, err)
+
+	res, err := b.GetEvents(ctx, 4)
+	require.Equal(t, event3Data, rlp.RawValue(res[0].Data()))
+	require.NoError(t, err)
+
+	err = b.Unwind(ctx, &types.Header{Number: big.NewInt(3)})
+	require.NoError(t, err)
+
+	_, err = b.GetEvents(ctx, 4)
+	require.Error(t, err)
+
+	cancel()
+	wg.Wait()
+}

--- a/polygon/bridge/db.go
+++ b/polygon/bridge/db.go
@@ -4,6 +4,8 @@ import (
 	"bytes"
 	"context"
 	"encoding/binary"
+	"errors"
+	"fmt"
 	"time"
 
 	"github.com/ledgerwatch/erigon-lib/kv"
@@ -11,6 +13,19 @@ import (
 	"github.com/ledgerwatch/erigon/polygon/heimdall"
 	"github.com/ledgerwatch/erigon/polygon/polygoncommon"
 )
+
+/*
+	BorEventNums stores the last event ID of the last sprint.
+
+	e.g. For block 10 with events [1,2,3], block 15 with events [4,5,6] and block 20 with events [7,8].
+	The DB will have the following.
+		10: 0 (initialized at zero, NOTE: Polygon does not have and event 0)
+		15: 3
+		20: 6
+
+	To get the events for block 15, we look up the map for 15 and 20 and get back 3 and 6. So our
+	ID range is [4,6].
+*/
 
 var databaseTablesCfg = kv.TableCfg{
 	kv.BorEvents:    {},
@@ -109,10 +124,6 @@ func (s *MdbxStore) GetSprintLastEventID(ctx context.Context, lastID uint64, tim
 
 	kLastID := make([]byte, 8)
 	binary.BigEndian.PutUint64(kLastID, lastID)
-
-	if bytes.Equal(kLastID, kDBLast) {
-		return lastID, nil
-	}
 
 	_, _, err = cursor.Seek(kLastID)
 	if err != nil {
@@ -225,6 +236,9 @@ func (s *MdbxStore) StoreEventID(ctx context.Context, eventMap map[uint64]uint64
 	return tx.Commit()
 }
 
+// GetEventIDRange returns the state sync event ID range for the given block number.
+// An error is thrown if the block number is not found in the database. If the given block
+// number is the last in the database, then the second uint64 (representing end ID) is 0.
 func (s *MdbxStore) GetEventIDRange(ctx context.Context, blockNum uint64) (uint64, uint64, error) {
 	var start, end uint64
 
@@ -237,22 +251,31 @@ func (s *MdbxStore) GetEventIDRange(ctx context.Context, blockNum uint64) (uint6
 	kByte := make([]byte, 8)
 	binary.BigEndian.PutUint64(kByte, blockNum)
 
-	it, err := tx.RangeAscend(kv.BorEventNums, kByte, nil, 2)
+	cursor, err := tx.Cursor(kv.BorEventNums)
 	if err != nil {
 		return start, end, err
 	}
 
-	for it.HasNext() {
-		_, v, err := it.Next()
-		if err != nil {
-			return start, end, err
-		}
+	_, v, err := cursor.SeekExact(kByte)
+	if err != nil {
+		return start, end, err
+	}
+	if v == nil { // we don't have a map
+		return start, end, errors.New(fmt.Sprintf("map not available for block %d", blockNum))
+	}
 
-		if start == 0 {
-			err = binary.Read(bytes.NewReader(v), binary.BigEndian, &start)
-		} else {
-			err = binary.Read(bytes.NewReader(v), binary.BigEndian, &end)
-		}
+	err = binary.Read(bytes.NewReader(v), binary.BigEndian, &start)
+	if err != nil {
+		return start, end, err
+	}
+
+	_, v, err = cursor.Next()
+	if err != nil {
+		return start, end, err
+	}
+
+	if v != nil { // may be empty if blockNum is the last entry
+		err = binary.Read(bytes.NewReader(v), binary.BigEndian, &end)
 		if err != nil {
 			return start, end, err
 		}
@@ -271,22 +294,21 @@ func (s *MdbxStore) PruneEventIDs(ctx context.Context, blockNum uint64) error {
 	kByte := make([]byte, 8)
 	binary.BigEndian.PutUint64(kByte, blockNum)
 
-	it, err := tx.RangeDescend(kv.BorEvents, nil, kByte, 0)
+	cursor, err := tx.Cursor(kv.BorEventNums)
+	if err != nil {
+		return err
+	}
+	defer cursor.Close()
+
+	var k []byte
+	for k, _, err = cursor.Seek(kByte); err == nil && k != nil; k, _, err = cursor.Next() {
+		if err := tx.Delete(kv.BorEventNums, k); err != nil {
+			return err
+		}
+	}
 	if err != nil {
 		return err
 	}
 
-	for it.HasNext() {
-		k, _, err := it.Next()
-		if err != nil {
-			return err
-		}
-
-		err = tx.Delete(kv.BorEventNums, k)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
+	return tx.Commit()
 }

